server/modules/routing/dbseerroute/dbseerroute.c

```diff
@@ -0,0 +1,1181 @@
+            rc = backend_dcb->func.write(backend_dcb, queue);
+            break;
+    }
+
+    MXS_INFO("Routed [%s] to '%s'%s%s",
+             STRPACKETTYPE(mysql_command),
+             backend_dcb->server->unique_name,
+             trc ? ": " : ".",
+             trc ? trc : "");
+    free(trc);
+
+return_rc:
+
+    return rc;
+}
+
+/**
+ * Display router diagnostics
+ *
+ * @param instance	Instance of the router
+ * @param dcb		DCB to send diagnostics to
+ */
+static void
+diagnostics(ROUTER *router, DCB *dcb)
+{
+    ROUTER_INSTANCE *router_inst = (ROUTER_INSTANCE *) router;
+    char *weightby;
+
+    dcb_printf(dcb, "\tNumber of router sessions:   	%d\n",
+               router_inst->stats.n_sessions);
+    dcb_printf(dcb, "\tCurrent no. of router sessions:	%d\n",
+               router_inst->service->stats.n_current);
+    dcb_printf(dcb, "\tNumber of queries forwarded:   	%d\n",
+               router_inst->stats.n_queries);
+    if ((weightby = serviceGetWeightingParameter(router_inst->service))
+        != NULL)
+    {
+        dcb_printf(dcb, "\tConnection distribution based on %s "
+                   "server parameter.\n",
+                   weightby);
+        dcb_printf(dcb,
+                   "\t\tServer               Target %% Connections\n");
+        for (SERVER_REF *ref = router_inst->service->dbref; ref; ref = ref->next)
+        {
+            dcb_printf(dcb, "\t\t%-20s %3.1f%%     %d\n",
+                       ref->server->unique_name,
+                       (float) ref->weight / 10,
+                       ref->connections);
+        }
+    }
+}
+
+/**
+ * Client Reply routine
+ *
+ * The routine will reply to client data from backend server
+ *
+ * @param       instance        The router instance
+ * @param       router_session  The router session
+ * @param       backend_dcb     The backend DCB
+ * @param       queue           The GWBUF with reply data
+ */
+static void
+clientReply(ROUTER *instance, void *router_session, GWBUF *queue, DCB *backend_dcb)
+{
+    /* perform logging if log file has been specified.*/
+    ROUTER_INSTANCE *inst = (ROUTER_INSTANCE *) instance;
+    ROUTER_CLIENT_SES *router_cli_ses = (ROUTER_CLIENT_SES *) router_session;
+    if (inst->log_filename)
+    {
+        struct timeval tv, diff;
+        int i;
+
+        /* found 'commit' and sql statements exists.*/
+        if (router_cli_ses->sql_end && router_cli_ses->sql_index > 0)
+        {
+            gettimeofday(&tv, NULL);
+            timersub(&tv, &(router_cli_ses->current_start), &diff);
+
+            /* get latency.*/
+            uint64_t millis = (diff.tv_sec * (uint64_t)1000 + diff.tv_usec / 1000);
+            /* get timestamp.*/
+            uint64_t timestamp = (tv.tv_sec + (tv.tv_usec / (1000*1000)));
+            *(router_cli_ses->sql + router_cli_ses->sql_index) = '\0';
+
+            char *delimiter = inst->log_delimiter;
+            char* server_uniquename = router_cli_ses->backend->server->unique_name;
```

Comment:
I think this could be done in the tpmfilter as well if we add a `SERVER` pointer to the GWBUF which tells where this buffer came from. It would allow filters to track the origin of a particular buffer.